HW2/P5/physics.pyx

```diff
@@ -113,18 +134,27 @@ cpdef update(FLOAT[:, ::1] XY,
         #
         # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
         # scheduling.
-        for i in range(count):
-            sub_update(XY, V, R, i, count, Grid, grid_spacing)
+        for i in prange(count, num_threads=4, schedule='static',chunksize=count/4):
+            sub_update(XY, V, R, i, count, Grid, grid_spacing, grid_size, locks)
 
         # update positions
         #
         # SUBPROBLEM 1: parallelize this loop over 4 threads (with static
         #    scheduling).
         # SUBPROBLEM 2: update the grid values.
-        for i in range(count):
+        for i in prange(count, num_threads=4, schedule='static', chunksize=count/4):
+            # Must be careful about grid updates: don't want to do any out of bounds accesses for things which organically
+            # left.
+            if (<int> (XY[i, 0]/grid_spacing) < grid_size) and (<int> (XY[i, 1]/grid_spacing) < grid_size) and (<int> (XY[i, 1]/grid_spacing) >= 0) and (<int> (XY[i, 0]/grid_spacing)) >= 0:
+                Grid[<int> (XY[i, 0]/grid_spacing), <int> (XY[i, 1]/grid_spacing)] = -1
+            
+            # Actually update the positions.
             for dim in range(2):
                 XY[i, dim] += V[i, dim] * t
 
+            # Tell the grid where the new ball is.
+            if (<int> (XY[i, 0]/grid_spacing) < grid_size) and (<int> (XY[i, 1]/grid_spacing) < grid_size) and (<int> (XY[i, 1]/grid_spacing) >= 0) and (<int> (XY[i, 0]/grid_spacing)) >= 0:
+                Grid[<int> (XY[i, 0]/grid_spacing), <int> (XY[i, 1]/grid_spacing)] = i
 
 def preallocate_locks(num_locks):
     cdef omp_lock_t *locks = get_N_locks(num_locks)
```

Comment:
Generally nice. would be better if there's a graph to illustrate the efficiency increase.